SimMuon/GEMDigitizer/src/GEMSignalModel.cc

```diff
@@ -0,0 +1,151 @@
+#include "SimMuon/GEMDigitizer/interface/GEMSignalModel.h"
+#include "Geometry/GEMGeometry/interface/GEMEtaPartitionSpecs.h"
+#include "Geometry/CommonTopologies/interface/TrapezoidalStripTopology.h"
+#include "Geometry/GEMGeometry/interface/GEMGeometry.h"
+#include "FWCore/Utilities/interface/Exception.h"
+#include "CLHEP/Random/RandFlat.h"
+#include "CLHEP/Random/RandPoissonQ.h"
+#include "CLHEP/Random/RandGaussQ.h"
+#include <cmath>
+#include <utility>
+#include <map>
+
+GEMSignalModel::GEMSignalModel(const edm::ParameterSet& config)
+    : GEMDigiModel(config),
+      averageEfficiency_(config.getParameter<double>("averageEfficiency")),
+      averageShapingTime_(config.getParameter<double>("averageShapingTime")),
+      timeResolution_(config.getParameter<double>("timeResolution")),
+      timeJitter_(config.getParameter<double>("timeJitter")),
+      signalPropagationSpeed_(config.getParameter<double>("signalPropagationSpeed")),
+      digitizeOnlyMuons_(config.getParameter<bool>("digitizeOnlyMuons")),
+      resolutionX_(config.getParameter<double>("resolutionX")) {}
+
+GEMSignalModel::~GEMSignalModel() {}
+
+void GEMSignalModel::simulate(const GEMEtaPartition* roll,
+                              const edm::PSimHitContainer& simHits,
+                              CLHEP::HepRandomEngine* engine,
+                              Strips& strips_,
+                              DetectorHitMap& detectorHitMap_) {
+  bool digiMuon = false;
+  bool digiElec = false;
+  for (const auto& hit : simHits) {
+    if (std::abs(hit.particleType()) != 13 && digitizeOnlyMuons_)
+      continue;
+    double elecEff = 0.;
+    double partMom = hit.pabs();
+    double checkMuonEff = CLHEP::RandFlat::shoot(engine, 0., 1.);
+    double checkElecEff = CLHEP::RandFlat::shoot(engine, 0., 1.);
+    if (std::abs(hit.particleType()) == 13 && checkMuonEff < averageEfficiency_)
+      digiMuon = true;
+    if (std::abs(hit.particleType()) != 13)  //consider all non muon particles with gem efficiency to electrons
+    {
+      if (partMom <= 1.95e-03)
```

Comment:
lots of magic numbers here, should all be named constants (or configurable?)

Reply:
Magic numbers are displaced to named constants.